### PR TITLE
rustc_typeck: use subtyping on the LHS of binops.

### LINKED ISSUE #45425: Binary operator LHS type is fixed instead of being subtyped
This started with a [Stack Overflow question](https://stackoverflow.com/q/46857955/155423) which boiled down to:

```rust
use std::ops::Rem;

fn powm<T>(fbase: &T, modulus: &T)
where
    for<'x> &'x T: Rem<&'x T, Output = T>,
{
    fbase % modulus;
}

fn main() {}
```

<details>
<summary>Complete error</summary>

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/main.rs:7:11
  |
7 |     fbase % modulus;
  |           ^
  |
note: first, the lifetime cannot outlive the lifetime 'b as defined on the function body at 3:1...
 --> src/main.rs:3:1
  |
3 | / fn powm<'a, 'b, T>(fbase: &'a T, modulus: &'b T)
4 | | where
5 | |     for<'x> &'x T: Rem<&'x T, Output = T>,
6 | | {
7 | |     fbase % modulus;
8 | | }
  | |_^
note: ...so that reference does not outlive borrowed content
 --> src/main.rs:7:13
  |
7 |     fbase % modulus;
  |             ^^^^^^^
note: but, the lifetime must be valid for the lifetime 'a as defined on the function body at 3:1...
 --> src/main.rs:3:1
  |
3 | / fn powm<'a, 'b, T>(fbase: &'a T, modulus: &'b T)
4 | | where
5 | |     for<'x> &'x T: Rem<&'x T, Output = T>,
6 | | {
7 | |     fbase % modulus;
8 | | }
  | |_^
note: ...so that types are compatible (expected std::ops::Rem, found std::ops::Rem<&T>)
 --> src/main.rs:7:11
  |
7 |     fbase % modulus;
  |           ^
```

</details>

----

Some poking around found some workarounds:

 - If we say that the input references can be unified to the same
   lifetime, it works:
   
    ```rust
    fn powm<'a, T>(fbase: &'a T, modulus: &'a T)
    ```

 - If we say that `'b` outlives `'a`, it works:

    ```rust
    fn powm<'a, 'b: 'a, T>(fbase: &'a T, modulus: &'b T)
    ```
   
 - If we say that we can have two different lifetimes in the operator, it works:
   
    ```rust
    for<'x, 'y> &'x T: Rem<&'y T, Output = T>,
    ```

 - If we directly call the `Rem::rem` method, it works:
   
    ```rust
    Rem::rem(fbase, modulus);
   ```

 - If we dereference and re-reference, it works:
   
    ```rust
    &*fbase % &*modulus;
    ```

I poked @eddyb for further explanation why the original case didn't work. Our tidied IRC conversation:

> @eddyb 
> The answer is `Rem::rem(fbase, modulus);` *doesn't* do unification, it does a reborrow coercion.
> That is, it's equivalent to `&*fbase % &*modulus`
> You can figure this out by having an operator implemented with `&str` on the RHS and passing `&String` or other stuff like that.
> If it works with `&String`, then this is generally approved

> @shepmaster
> But what's the difference of `a % b` and `Rem::rem(a, b)` — I thought that was straight desugaring?

> @eddyb 
> There is no such thing as a straight desugaring; the first is `Rem::rem(a, b)` the second is `(Rem::rem)(coerce a, coerce b)`
> IDK, I think we should coerce, but we can't coerce the LHS, because......... what type to coerce to?
> It has to be asymmetrical. "has to"
> Wait, no, why is this a... I might be wrong; yeah no I am wrong.

> @eddyb 
> This is about traits... so `Rem::rem(a, b)` is `<_ as Rem<_>>::rem(a, b)` and `a % b` is `<typeof a as Rem<typeof b>>::rem(a, b)`
> So we do coercions on the RHS, but *I think* we need to also subtype the LHS, which we're not doing right now.

> @shepmaster
> There's some piece of "this doesn't work because the lifetimes are disjoint, even though they could be unified at least as much as the lifetime of the function call"

> @eddyb 
> It's not really disjointness; hint: an asymmetrical `'b: 'a` or whatever you added only helps because the RHS is getting coerced. The RHS can be a subtype but the LHS is fixed.
> Lemme rephrase :D
> `<typeof a as Rem<_>>::rem(a, b)` whereas you want `_ extends (typeof a)` or something as the `Self` for `Rem`

> @eddyb 
> We already use forced subtyping to have more flexibility around unsizing coercions because the trait matching of `T: Unsize<U>` is invariant.
> It wasn't invariant for a while, which was a soundness bug..... introduced by yours truly in the original implementation.

> @eddyb 
> [This shows `lhs_ty` being used for `Self` and `rhs_ty_var` for the trait parameters](https://github.com/rust-lang/rust/blob/b633341c49498488b6e3fce501e23e9f9f8fadec/src/librustc_typeck/check/op.rs#L178-L189) so the LHS type is fixed instead of being subtyped.

> @eddyb 
> [This is where we do the right thing](https://github.com/rust-lang/rust/blob/b633341c49498488b6e3fce501e23e9f9f8fadec/src/librustc_typeck/check/coercion.rs#L495-L501), going through an inference variable & extra subtyping (or LUB but that doesn't apply here) to make something more flexible than what the trait system allows.

### PR DESCRIPTION
**Pull request by eddyb**
Fixes #45425.

r? @nikomatsakis

### CODE REVIEW

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/b92d8f274ebaf59b96fce0e2459ca76d) has only false positives.

File: src/librustc_typeck/check/coercion.rs [hunk 1]
```diff
@@ -763,7 +763,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
         let (adjustments, _) = self.register_infer_ok_obligations(ok);
         self.apply_adjustments(expr, adjustments);
-        Ok(target)
+        Ok(self.resolve_type_vars_with_obligations(target))
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/coercion.rs]: Is there a reason for this? Just to improve debug output, or something else?

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -88,10 +94,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     pub fn demand_coerce_diag(&self,
                               expr: &hir::Expr,
                               checked_ty: Ty<'tcx>,
-                              expected: Ty<'tcx>) -> Option<DiagnosticBuilder<'tcx>> {
+                              expected: Ty<'tcx>)
+                              -> (Ty<'tcx>, Option<DiagnosticBuilder<'tcx>>) {
         let expected = self.resolve_type_vars_with_obligations(expected);
 
-        if let Err(e) = self.try_coerce(expr, checked_ty, self.diverges.get(), expected) {
+        self.try_coerce(expr, checked_ty, self.diverges.get(), expected).map_err(|e| {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/demand.rs]: Nit: this whole `foo.map_err().map().unwrap_or` is getting a bit over the top and kind of hard to follow. Can we just rewrite this with a `match`?

Something like:

```rust
let e = match self.try_coerce(...) {
    Ok(ty) => return (ty, None),
    Err(e) => e,
};

...

(expected, some(err))
```

File: src/librustc_typeck/check/op.rs [hunk 3]
```diff
@@ -164,17 +156,29 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+        let lhs_pref = match is_assign {
+            IsAssign::Yes => PreferMutLvalue,
+            IsAssign::No => NoPreference
+        };
+        // Find a suitable supertype of the LHS expression's type, by coercing to
+        // a type variable, to pass as the `Self` to the trait, avoiding invariant
+        // trait matching creating lifetime constraints that are too strict.
+        // E.g. adding `&'a T` and `&'b T`, given `&'x T: Add<&'x T>`, will result
+        // in `&'a T <: &'x T` and `&'b T <: &'x T`, instead of `'a = 'b = 'x`.
+        let lhs_ty = self.check_expr_coercable_to_type_with_lvalue_pref(lhs_expr,
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/op.rs]: Why use *coercion* here rather than subtyping? I guess it is equivalent anyway, at least for the way we handle coercions now, since if you coerce with a target type of a variable you just get subtyping anyway, right? 

OTOH, if we ever support "deferred coercion", I don't see any reason *not* to use coercion here, at least for by-value operators like `+` -- not entirely sure about `==`.

File: src/test/run-pass/issue-32008.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+impl<'a> Add<&'a Foo> for &'a Foo {
+    type Output = ();
+    fn add(self, rhs: &'a Foo) {}
+}
+
+fn try_to_add(input: &Foo) {
+    let local = Foo;
+    &*input + &local; // ok
+    input + &local; // Error!
```
Comment (nikomatsakis) [inline src/test/run-pass/issue-32008.rs]: Can you remove the `// Error!` comment

File: src/test/run-pass/issue-32008.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::ops::Add;
+
+struct Foo;
```
Comment (nikomatsakis) [inline src/test/run-pass/issue-32008.rs]: Can you add a comment explaining what this is all about? i.e,. "Check that we permit subtyping for operator LHS" or something.

**Review by nikomatsakis — COMMENTED** [review]
Seems ok for the most part. I left a few questions to make sure I'm following along.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_typeck/check/coercion.rs]: It's used to avoid repeatedly doing this in the callers.

File: src/librustc_typeck/check/demand.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_typeck/check/demand.rs]: I wanted to avoid intending the whole function right, but I'm okay with a match either.

File: src/librustc_typeck/check/op.rs (hunk 3 above)
Comment (eddyb) [reply src/librustc_typeck/check/op.rs]: Subtyping is trickier to get right, I'm far more comfortable using coercion instead.

File: src/test/run-pass/issue-32008.rs (hunk 4 above)
Comment (eddyb) [reply src/test/run-pass/issue-32008.rs]: Heh I was considering it.

File: src/librustc_typeck/check/op.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/op.rs]: Trickier in what sense? Coercing is a superset of subtyping, so using it can only cause more things to happen.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/coercion.rs]: So you mean that the callers rely on type variables having been resolved in the return value?

Comment (eddyb) [reply src/librustc_typeck/check/coercion.rs]: Yeah, in the changed binop code.

File: src/librustc_typeck/check/op.rs (hunk 3 above)
Comment (eddyb) [reply src/librustc_typeck/check/op.rs]: I mean doing subtyping manually - I know coercion uses subtyping correctly and I'd rather use an API where it's obvious how not to misuse it.

**Comment by eddyb** [comment]
@nikomatsakis Let me know if you want more changes to this.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/coercion.rs]: Then I think we should add some comment that this is important, and also comment the interface of the coercion function. I'm not all that keen on having such a guarantee, though -- it's a subtle link between code imo -- but if you can at least document who cares about it, it would be helpful.

File: src/librustc_typeck/check/op.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/op.rs]: I guess I'm just nervous that, if coercion gets smarter, we'll now be performing random coercions here -- as opposed to just subtyping -- which maybe we do not want to do. But I can't really think of a good reason *not* to use coercion.

Comment (eddyb) [reply src/librustc_typeck/check/op.rs]: We already coerce the RHS FWIW.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_typeck/check/coercion.rs]: Well then I'll just throw away most of the changes, I guess... I thought this API was much nicer but I can just as well have two extra lines in `op.rs` instead.

File: src/librustc_typeck/check/op.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/op.rs]: That does make me feel better =)

**Comment by nikomatsakis** [comment]
@eddyb are you going to either comment or remove the "resolve type variables" line? I'm ok either way. Other than that, r=me.

**Comment by alexcrichton** [comment]
It looks like this [may have regressed](https://perf.rust-lang.org/?start=2017-10-29&end=2017-11-02&stat=instructions%3Au&absolute=false) the `inflate` benchmark by ~5%?

**Comment by eddyb** [comment]
I suppose that more inference variables are created even without any lifetimes being involved.

@nikomatsakis, are there any plans to handle lifetime subtyping/generalization more efficiently during typeck?
